pkg/rcmanager/ipinstance.go

```diff
@@ -0,0 +1,196 @@
+/*
+  Copyright 2021 The Hybridnet Authors.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package rcmanager
+
+import (
+	"context"
+	"fmt"
+
+	networkingv1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/constants"
+	"github.com/oecp/rama/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+)
+
+// reconcile one single node, update/add/remove everything about it's
+// corresponding remote vtep.
+func (m *Manager) reconcileIPInstance(nodeName string) error {
+	klog.Infof("[remote cluster] Starting reconcile ipinstance from cluster %v, node name=%v", m.ClusterName, nodeName)
+	if len(nodeName) == 0 {
+		return nil
+	}
+	var (
+		node     *v1.Node
+		err      error
+		vtepName string
+	)
+	vtepName = utils.GenRemoteVtepName(m.ClusterName, nodeName)
+	node, err = m.NodeLister.Get(nodeName)
+	if err != nil {
+		if k8serror.IsNotFound(err) {
+			err = m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Delete(context.TODO(), vtepName, metav1.DeleteOptions{})
+			if k8serror.IsNotFound(err) {
+				return nil
+			}
+		}
+		return err
+	}
+
+	var (
+		vtepIP     = node.Annotations[constants.AnnotationNodeVtepIP]
+		vtepMac    = node.Annotations[constants.AnnotationNodeVtepMac]
+		newVtep    = false
+		remoteVtep *networkingv1.RemoteVtep
+	)
+
+	remoteVtep, err = m.RemoteVtepLister.Get(vtepName)
+	if err != nil {
+		if !k8serror.IsNotFound(err) {
+			return err
+		}
+		remoteVtep = utils.NewRemoteVtep(m.ClusterName, m.RemoteClusterUID, vtepIP, vtepMac,
+			node.Annotations[constants.AnnotationNodeLocalVxlanIPList], node.Name, nil)
+		newVtep = true
+	}
+
+	var (
+		curTime = metav1.Now()
+		desired []string
+		// vtepIP or vtepMac changed
+		vtepChanged           bool
+		endpointILListChanged bool
+	)
+
+	desired, err = m.pickEndpointListFromNode(node)
+	if err != nil {
+		return err
+	}
+	endpointILListChanged = utils.DifferentSetFromStringSlice(remoteVtep.Spec.EndpointIPList, desired)
+
+	remoteVtep = remoteVtep.DeepCopy()
+	remoteVtep.Spec.VtepIP = vtepIP
+	remoteVtep.Spec.VtepMAC = vtepMac
+	remoteVtep.Spec.EndpointIPList = desired
+	remoteVtep.Status.LastModifyTime = curTime
+
+	vtepChanged = vtepIP != "" && vtepMac != "" && (vtepIP != remoteVtep.Spec.VtepIP || vtepMac != remoteVtep.Spec.VtepMAC)
+	if !newVtep && !endpointILListChanged && !vtepChanged {
```

Comment:
Maybe `node.Annotations[constants.AnnotationNodeLocalVxlanIPList]` changed should be count on `vtepChanged`, just link `vtepIP` and `vtepIP` are changed.